rfcs/0126-generic-worker-v2.md

```diff
@@ -0,0 +1,206 @@
+# RFC 126 - Megalodon worker (a.k.a. generic-worker v2)
+* Comments: [#126](https://api.github.com/repos/taskcluster/taskcluster-rfcs/issues/126)
+* Proposed by: @walac
+
+# Summary
+
+The worker is the component of Taskcluster responsible to execute the
+job described in the task payload. It runs on a cloud instance or
+data center server and picks its jobs from the
+[queue](https://github.com/taskcluster/taskcluster-queue).
+
+Taskcluster projects currently provides two workers:
+
+* [docker-worker](https://github.com/taskcluster/docker-worker): executes
+Taskcluster jobs inside a docker container.
+
+* [generic-worker](https://github.com/taskcluster/generic-worker): executes
+Taskcluster jobs natively in the running machine.
+
+Although workers may run tasks in rather different environments, they share
+a fair amount of features, and it is reasonable to think they could share
+a common code base. Indeed this was the finality of
+[taskcluster-worker](https://github.com/taskcluster/taskcluster-worker),
+a worker aimed to support multiple execution environments. For reasons
+explained later in this document, it failed to achieve this goal.
+
+# Motivation
+
+Both docker-worker and generic-worker were designed to accomplish a very
+specific task and do it well, but their architecture opens little room
+to support different execution environments, hence the proposal for a
+ground up worker.
+
+The initial thought was to extend generic-worker to support docker-worker
+payloads. This way we could benefit from the stability of generic-worker
+code base. The point is that any sane implementation of docker-worker
+payload support would lead to such a heavy code change that
+the "code stability" claim would no longer be true; given that
+generic-worker architecture was never thought to support
+multiple payloads, like taskcluster-worker did.
```

Comment:
How different are the payloads? I believe there's a single location in tree that sets the payload for these tasks, so it's possible this RFC is bending over backwards to avoid a one-time hit in the in-tree task definitions.